src/OpenID/Connect/Client/Flow/AuthorizationCode.hs

```diff
@@ -399,7 +399,7 @@ exchangeCodeForIdentityToken https now disco creds user = do
       req <- maybe
         (throwError (InvalidProviderTokenEndpointError (uriToText (getURI uri)))) pure
         (requestFromURI (Right (getURI uri)))
-      applyRequestAuthentication creds authMethods
```

Comment:
The `authenticationSuccess` and `exchangeCodeForIdentityToken` functions should be updated to use the new type class.  However, the multi-parameter type class `Authenticatable` won't work here because there's no way (that I can see) to select the correct instance given a `Discovery` document and a `Credentials` record.

Therefore, the `authenticationSuccess` function doesn't work with any authentication extensions and fails to advertise that it will silently ignore them.  This leads me to believe that the current approach isn't going to work.

We need a way for `authenticationSuccess` to indicate that authentication extensions must be dealt with.  We might need a new type and function to correctly represent the authentication being requested and the credentials that are available.  Does that make sense?

Reply:
what about making authenticationSuccess and exchangeCodeForIdentityToken functions - members of typeclass?

Reply:
I'm not sure how a type class helps here.  Can you explain further?

I'm not necessarily against using a type class, but the general wisdom of the Haskell community is that type classes that don't have associated laws should be a last resort.  It's pretty easy to replace a type class with either a single function or a record of functions.